docs/new/node-setup/configuration.md


# Configuration Settings

Every Planetmint Server configuration setting has two names: a config-file name and an environment variable name. For example, one of the settings has the config-file name `database.host` and the environment variable name `PLANETMINT_DATABASE_HOST`. Here are some more examples:

`database.port` ↔ `PLANETMINT_DATABASE_PORT`

`database.keyfile_passphrase` ↔ `PLANETMINT_DATABASE_KEYFILE_PASSPHRASE`

`server.bind` ↔ `PLANETMINT_SERVER_BIND`

The value of each setting is determined according to the following rules:

* If it's set by an environment variable, then use that value
* Otherwise, if it's set in a local config file, then use that value
* Otherwise, use the default value

The local config file is `$HOME/.planetmint` by default (a file which might not even exist), but you can tell Planetmint to use a different file by using the `-c` command-line option, e.g. `planetmint -c path/to/config_file.json start`
or using the `PLANETMINT_CONFIG_PATH` environment variable, e.g. `PLANETMINT_CONFIG_PATH=.my_planetmint_config planetmint start`.
Note that the `-c` command line option will always take precedence if both the `PLANETMINT_CONFIG_PATH` and the `-c` command line option are used.

You can read the current default values in the file [planetmint/\_\_init\_\_.py](https://github.com/planetmint/planetmint/blob/master/planetmint/__init__.py). (The link is to the latest version.)


## database.*

The settings with names of the form `database.*` are for the backend database
(currently only Tarantool). They are:

* `database.backend` can only be `localtarantool`, currently.
* `database.host` is the hostname (FQDN) of the backend database.
* `database.port` is self-explanatory.
* `database.user` is a user-chosen name for the database inside Tarantool, e.g. `planetmint`.
* `database.pass` is the password of the user for connection to tarantool listener.

There are two ways for Planetmint Server to authenticate itself with Tarantool (or a specific Tarantool service): no authentication, username/password.

**No Authentication**

If you use all the default Planetmint configuration settings, then no authentication will be used.

**Username/Password Authentication**

To use username/password authentication, a Tarantool instance must already be running somewhere (maybe in another machine), it must already have a spaces for use by Planetmint, and that database must already have a "readWrite" user with associated username and password.

**Default values**

```js
"database": {
    "backend": "tarantool",
    "host": "localhost",
    "port": 3301,
    "username": null,
    "password": null
    
}
```

## server.*

`server.bind`, `server.loglevel` and `server.workers`
are settings for the [Gunicorn HTTP server](http://gunicorn.org/), which is used to serve the [HTTP client-server API](../connecting/http-client-server-api).

`server.bind` is where to bind the Gunicorn HTTP server socket. It's a string. It can be any valid value for [Gunicorn's bind setting](http://docs.gunicorn.org/en/stable/settings.html#bind). For example:

* If you want to allow IPv4 connections from anyone, on port 9984, use `0.0.0.0:9984`
* If you want to allow IPv6 connections from anyone, on port 9984, use `[::]:9984`

In a production setting, we recommend you use Gunicorn behind a reverse proxy server such as NGINX. If Gunicorn and the reverse proxy are running on the same machine, then you can use `localhost:9984` (the default value), meaning Gunicorn will talk to the reverse proxy on port 9984. The reverse proxy could then be bound to port 80 (for HTTP) or port 443 (for HTTPS), so that external clients would connect using that port. For example:

[External clients]---(port 443)---[NGINX]---(port 9984)---[Gunicorn / Planetmint Server]

If Gunicorn and the reverse proxy are running on different machines, then `server.bind` should be `hostname:9984`, where hostname is the IP address or [FQDN](https://en.wikipedia.org/wiki/Fully_qualified_domain_name) of the reverse proxy.

There's [more information about deploying behind a reverse proxy in the Gunicorn documentation](http://docs.gunicorn.org/en/stable/deploy.html). (They call it a proxy.)

`server.loglevel` sets the log level of Gunicorn's Error log outputs. See
[Gunicorn's documentation](http://docs.gunicorn.org/en/latest/settings.html#loglevel)
for more information.

`server.workers` is [the number of worker processes](http://docs.gunicorn.org/en/stable/settings.html#workers) for handling requests. If set to `None`, the value will be (2 × cpu_count + 1). Each worker process has a single thread. The HTTP server will be able to handle `server.workers` requests simultaneously.

**Example using environment variables**

```text
export PLANETMINT_SERVER_BIND=0.0.0.0:9984
export PLANETMINT_SERVER_LOGLEVEL=debug
export PLANETMINT_SERVER_WORKERS=5
```

**Example config file snippet**

```js
"server": {
    "bind": "0.0.0.0:9984",
    "loglevel": "debug",
    "workers": 5,
}
```

**Default values (from a config file)**

```js
"server": {
    "bind": "localhost:9984",
    "loglevel": "info",
    "workers": null,
}
```

## wsserver.*


### wsserver.scheme, wsserver.host and wsserver.port

These settings are for the
[aiohttp server](https://aiohttp.readthedocs.io/en/stable/index.html),
which is used to serve the
[WebSocket Event Stream API](../connecting/websocket-event-stream-api).
`wsserver.scheme` should be either `"ws"` or `"wss"`
(but setting it to `"wss"` does *not* enable SSL/TLS).
`wsserver.host` is where to bind the aiohttp server socket and
`wsserver.port` is the corresponding port.
If you want to allow connections from anyone, on port 9985,
set `wsserver.host` to 0.0.0.0 and `wsserver.port` to 9985.

**Example using environment variables**

```text
export PLANETMINT_WSSERVER_SCHEME=ws
export PLANETMINT_WSSERVER_HOST=0.0.0.0
export PLANETMINT_WSSERVER_PORT=9985
```

**Example config file snippet**

```js
"wsserver": {
    "scheme": "wss",
    "host": "0.0.0.0",
    "port": 65000
}
```

**Default values (from a config file)**

```js
"wsserver": {
    "scheme": "ws",
    "host": "localhost",
    "port": 9985
}
```

### wsserver.advertised_scheme, wsserver.advertised_host and wsserver.advertised_port

These settings are for the advertising the Websocket URL to external clients in
the root API endpoint. These configurations might be useful if your deployment
is hosted behind a firewall, NAT, etc. where the exposed public IP or domain is
different from where Planetmint is running.

**Example using environment variables**

```text
export PLANETMINT_WSSERVER_ADVERTISED_SCHEME=wss
export PLANETMINT_WSSERVER_ADVERTISED_HOST=myplanetmint.io
export PLANETMINT_WSSERVER_ADVERTISED_PORT=443
```

**Example config file snippet**

```js
"wsserver": {
    "advertised_scheme": "wss",
    "advertised_host": "myplanetmint.io",
    "advertised_port": 443
}
```

**Default values (from a config file)**

```js
"wsserver": {
    "advertised_scheme": "ws",
    "advertised_host": "localhost",
    "advertised_port": 9985
}
```

## log.*

The `log.*` settings are to configure logging.

**Example**

```js
{
    "log": {
        "file": "/var/log/planetmint.log",
        "error_file": "/var/log/planetmint-errors.log",
        "level_console": "info",
        "level_logfile": "info",
        "datefmt_console": "%Y-%m-%d %H:%M:%S",
        "datefmt_logfile": "%Y-%m-%d %H:%M:%S",
        "fmt_console": "%(asctime)s [%(levelname)s] (%(name)s) %(message)s",
        "fmt_logfile": "%(asctime)s [%(levelname)s] (%(name)s) %(message)s",
        "granular_levels": {}
}
```

**Default values**

```js
{
    "log": {
        "file": "~/planetmint.log",
        "error_file": "~/planetmint-errors.log",
        "level_console": "info",
        "level_logfile": "info",
        "datefmt_console": "%Y-%m-%d %H:%M:%S",
        "datefmt_logfile": "%Y-%m-%d %H:%M:%S",
        "fmt_logfile": "[%(asctime)s] [%(levelname)s] (%(name)s) %(message)s (%(processName)-10s - pid: %(process)d)",
        "fmt_console": "[%(asctime)s] [%(levelname)s] (%(name)s) %(message)s (%(processName)-10s - pid: %(process)d)",
        "granular_levels": {}
}
```

### log.file

The full path to the file where logs should be written.
The user running `planetmint` must have write access to the
specified path.

**Log rotation:** Log files have a size limit of about 200 MB
and will be rotated up to five times.
For example, if `log.file` is set to `"~/planetmint.log"`, then
logs would always be written to `planetmint.log`. Each time the file
`planetmint.log` reaches 200 MB it will be closed and renamed
`planetmint.log.1`. If `planetmint.log.1` and `planetmint.log.2` already exist they
would be renamed `planetmint.log.2` and `planetmint.log.3`. This pattern would be
applied up to `planetmint.log.5` after which `planetmint.log.5` would be
overwritten by `planetmint.log.4`, thus ending the rotation cycle of whatever
logs were in `planetmint.log.5`.

### log.error_file

Similar to `log.file` (see above), this is the
full path to the file where error logs should be written.

### log.level_console

The log level used to log to the console. Possible allowed values are the ones
defined by [Python](https://docs.python.org/3.9/library/logging.html#levels),
but case-insensitive for the sake of convenience:

```text
"critical", "error", "warning", "info", "debug", "notset"
```

### log.level_logfile

The log level used to log to the log file. Possible allowed values are the ones
defined by [Python](https://docs.python.org/3.9/library/logging.html#levels),
but case-insensitive for the sake of convenience:

```text
"critical", "error", "warning", "info", "debug", "notset"
```

### log.datefmt_console

The format string for the date/time portion of a message, when logged to the
console.

For more information on how to construct the format string please consult the
table under [Python's documentation of time.strftime(format[, t])](https://docs.python.org/3.9/library/time.html#time.strftime).

### log.datefmt_logfile

The format string for the date/time portion of a message, when logged to a log
 file.

For more information on how to construct the format string please consult the
table under [Python's documentation of time.strftime(format[, t])](https://docs.python.org/3.9/library/time.html#time.strftime).

### log.fmt_console

A string used to format the log messages when logged to the console.

For more information on possible formatting options please consult Python's
documentation on
[LogRecord attributes](https://docs.python.org/3.9/library/logging.html#logrecord-attributes).

### log.fmt_logfile

A string used to format the log messages when logged to a log file.

For more information on possible formatting options please consult Python's
documentation on
[LogRecord attributes](https://docs.python.org/3.9/library/logging.html#logrecord-attributes).

### log.granular_levels

Log levels for Planetmint's modules. This can be useful to control the log
level of specific parts of the application. As an example, if you wanted the
logging of the `core.py` module to be more verbose, you would set the
 configuration shown in the example below.

**Example**

```js
{
    "log": {
        "granular_levels": {
            "bichaindb.core": "debug"
        }
}
```

**Default value**

```js
{}
```

## tendermint.*

The settings with names of the form `tendermint.*` tell Planetmint Server
where it can connect to the node's Tendermint instance.

* `tendermint.host` is the hostname (FQDN)/IP address of the Tendermint instance.
* `tendermint.port` is self-explanatory.

**Example using environment variables**

```text
export PLANETMINT_TENDERMINT_HOST=tendermint
export PLANETMINT_TENDERMINT_PORT=26657
```Planetmint

**Default values**

```js
"tendermint": {
    "host": "localhost",
    "port": 26657
}
```


CHANGELOG.md
<!---
Copyright © 2020 Interplanetary Database Association e.V.,
Planetmint and IPDB software contributors.
SPDX-License-Identifier: (Apache-2.0 AND CC-BY-4.0)
Code is Apache-2.0 and docs are CC-BY-4.0
--->

# Change Log (Release Notes)

All _notable_ changes to this project will be documented in this file (`CHANGELOG.md`).

This project adheres to [the Python form of Semantic Versioning](https://packaging.python.org/tutorials/distributing-packages/#choosing-a-versioning-scheme) (or at least we try). The Planetmint public API _was_ defined in this file but that definition was moved and can now be found in [BEP-7](https://github.com/planetmint/BEPs/tree/master/7).

Contributors to this file, please follow the guidelines on [keepachangelog.com](http://keepachangelog.com/). Note that each version (or "release") is the name of a [Git _tag_](https://git-scm.com/book/en/v2/Git-Basics-Tagging) of a particular commit, so the associated date and time are the date and time of that commit (as reported by GitHub), _not_ the "Uploaded on" date listed on PyPI (which may differ).

For reference, the possible headings are:

* **Added** for new features.
* **Changed** for changes in existing functionality.
* **Deprecated** for once-stable features removed in upcoming releases.
* **Removed** for deprecated features removed in this release.
* **Fixed** for any bug fixes.
* **Security** to invite users to upgrade in case of vulnerabilities.
* **External Contributors** to list contributors outside of Planetmint GmbH.
* **Known Issues**
* **Notes**

## [Unreleased]

## [1.2.2] - 2022-23-09
* **Changed** A connection singleton got introduced to avoid connection issues and unstable states.



## [1.2.1] - 2022-20-09
* **Changed** Create model now validates for CID strings for asset["data"] and metadata
* **Changed** adjusted test cases

## [1.2.0] - 2022-09-05
* **Changed** disabled acceptance and integration tests, they have a circular dep. to the python driver
* **Changed** Metadata and asset["data"] types to string containing an IPLD hash
* **Fixed** Transaction generation bug that automatically assigned 'assets' to asset["data"]
* **Changed** adjusted test cases


## [1.1.0] - 2022-09-05
* **Changed** adjusted to zenroom calling convention of PRP #13 (breaking change)
* **Changed** zenroom test cases to comply to the new calling convention
* **Fixed** zenroom signing bug (call of wrong function)
* **Changed** using cryptoconditions 0.10.0 
* **Deprecated** usage of ripde160md as a address generation algorithm, isn't available from python 3.9.14 on, skipping these tests from now on.
* **Changed** script/ouptut tag to be of type array or object for schema v3.0 and v2.0 
* **Changed** added 'script' handling to the common/transactions.py class
* **Fixed** data input handling to the transaction fullfillment methods



## [1.0.1] - 2022-07-07
updated documentation

## [1.0.0] - 2022-07-05
### Feature Update
Tarantool integration

## [0.9.8] - 2022-06-27

### Feature Update
Changed license to AGPLv3


## [0.9.7] - 2022-06-17

### Feature Update
Deep Zenroom integration

## [0.9.6] - 2022-06-08

### Maintenance

* removed Korean documentation
* removed Korean and Chinese README

## [2.2.2] - 2020-08-12

### Security

Several dependencies updated including Flask that had vulnerability.

### Fixed

* Updated priv_validator key format in stack script (#2707)

### External Contributors

* @aostrun - [#2708](https://github.com/planetmint/planetmint/pull/2708)

## [2.2.1] - 2020-04-14

### Fixed

Gevent library API update is incompatible with  bigchaindb-abci 1.0.1 version.
Updated bigchaindb-abci.

## [2.2.0] - 2020-02-20

### Added

Support for multiple ABCI versions.

## [2.1.0] - 2019-11-06

### Added

Option for last transaction retrieval added.

## [2.0] - 2019-09-26

### Changed

Migrated from Tendermint 0.22.8 to 0.31.5.

## [2.0 Beta 9] - 2018-11-27

### Changed

Removed support for TLSv1 and TLSv1.1 in all NGINX config files. Kept support for TLSv1.2 and added support for TLSv1.3. [Pull Request #2601](https://github.com/planetmint/planetmint/pull/2601)

### Fixed

Fixed two issues with schema validation. Pull requests [#2606](https://github.com/planetmint/planetmint/pull/2606) & [#2607](https://github.com/planetmint/planetmint/pull/2607)

### External Contributors

[@gamjapark](https://github.com/gamjapark) and team translated all the [Planetmint root docs](https://docs.planetmint.com/en/latest/korean/index.html) into Korean. [Pull Request #2603](https://github.com/planetmint/planetmint/pull/2603)

## [2.0 Beta 8] - 2018-11-03

### Changed

* Revised the [Simple Deployment Template](http://docs.planetmint.com/projects/server/en/latest/simple-deployment-template/index.html) in the docs. Added NGINX to the mix. Pull Requests [#2578](https://github.com/planetmint/planetmint/pull/2578) and [#2579](https://github.com/planetmint/planetmint/pull/2579)
* Revised `nginx/nginx.conf` to enable CORS. [Pull Request #2580](https://github.com/planetmint/planetmint/pull/2580)

### Fixed

* Fixed a typo in the Kubernetes ConfigMap template. [Pull Request #2583](https://github.com/planetmint/planetmint/pull/2583)

### External Contributors

[@gamjapark](https://github.com/gamjapark) translated the main `README.md` file into Korean. [Pull Request #2592](https://github.com/planetmint/planetmint/pull/2592)

## [2.0 Beta 7] - 2018-09-28

Tag name: v2.0.0b7

### Added

Completed the implementation of chain-migration elections (BEP-42). Pull requests [#2553](https://github.com/planetmint/planetmint/pull/2553), [#2556](https://github.com/planetmint/planetmint/pull/2556), [#2558](https://github.com/planetmint/planetmint/pull/2558), [#2563](https://github.com/planetmint/planetmint/pull/2563) and [#2566](https://github.com/planetmint/planetmint/pull/2566)

### Changed


planetmint/config.py
import copy
import logging
import os

# from planetmint.log import DEFAULT_LOGGING_CONFIG as log_config
from planetmint.version import __version__  # noqa


class Singleton(type):
    _instances = {}

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]


class Config(metaclass=Singleton):
    def __init__(self):
        # from functools import reduce
        # PORT_NUMBER = reduce(lambda x, y: x * y, map(ord, 'Planetmint')) % 2**16
        # basically, the port number is 9984

        # The following variable is used by `planetmint configure` to
        # prompt the user for database values. We cannot rely on
        # _base_database_localmongodb.keys() because dicts are unordered.
        # I tried to configure
        self.log_config = DEFAULT_LOGGING_CONFIG
        db = "tarantool_db"
        self.__private_database_keys_map = {  # TODO Check if it is working after removing 'name' field
            "tarantool_db": ("host", "port"),
            "localmongodb": ("host", "port", "name"),
        }
        self.__private_database_localmongodb = {
            "backend": "localmongodb",
            "host": "localhost",
            "port": 27017,
            "name": "bigchain",
            "replicaset": None,
            "login": None,
            "password": None,
            "connection_timeout": 5000,
            "max_tries": 3,
            "ssl": False,
            "ca_cert": None,
            "certfile": None,
            "keyfile": None,
            "keyfile_passphrase": None,
            "crlfile": None,
        }
        self.__private_init_config = {
            "absolute_path": os.path.dirname(os.path.abspath(__file__)) + "/backend/tarantool/init.lua"
        }

        self.__private_drop_config = {
            "absolute_path": os.path.dirname(os.path.abspath(__file__)) + "/backend/tarantool/drop.lua"
        }
        self.__private_database_tarantool = {
            "backend": "tarantool_db",
            "connection_timeout": 5000,
            "max_tries": 3,
            "name": "universe",
            "reconnect_delay": 0.5,
            "host": "localhost",
            "port": 3303,
            "connect_now": True,
            "encoding": "utf-8",
            "login": "guest",
            "password": "",
            "service": "tarantoolctl connect",
            "init_config": self.__private_init_config,
            "drop_config": self.__private_drop_config,
        }

        self.__private_database_map = {
            "tarantool_db": self.__private_database_tarantool,
            "localmongodb": self.__private_database_localmongodb,
        }
        self.__private_config = {
            "server": {
                # Note: this section supports all the Gunicorn settings:
                #       - http://docs.gunicorn.org/en/stable/settings.html
                "bind": "localhost:9984",
                "loglevel": logging.getLevelName(self.log_config["handlers"]["console"]["level"]).lower(),
                "workers": None,  # if None, the value will be cpu_count * 2 + 1
            },
            "wsserver": {
                "scheme": "ws",
                "host": "localhost",
                "port": 9985,
                "advertised_scheme": "ws",
                "advertised_host": "localhost",
                "advertised_port": 9985,
            },
            "tendermint": {
                "host": "localhost",
                "port": 26657,
                "version": "v0.31.5",  # look for __tm_supported_versions__
            },
            "database": self.__private_database_map,
            "log": {
                "file": self.log_config["handlers"]["file"]["filename"],
                "error_file": self.log_config["handlers"]["errors"]["filename"],
                "level_console": logging.getLevelName(self.log_config["handlers"]["console"]["level"]).lower(),
                "level_logfile": logging.getLevelName(self.log_config["handlers"]["file"]["level"]).lower(),
                "datefmt_console": self.log_config["formatters"]["console"]["datefmt"],
                "datefmt_logfile": self.log_config["formatters"]["file"]["datefmt"],
                "fmt_console": self.log_config["formatters"]["console"]["format"],
                "fmt_logfile": self.log_config["formatters"]["file"]["format"],
                "granular_levels": {},
            },
        }
        self._private_real_config = copy.deepcopy(self.__private_config)
        # select the correct config defaults based on the backend
        self._private_real_config["database"] = self.__private_database_map[db]

    def init_config(self, db):
        self._private_real_config = copy.deepcopy(self.__private_config)
        # select the correct config defaults based on the backend
        self._private_real_config["database"] = self.__private_database_map[db]
        return self._private_real_config

    def get(self):
        return self._private_real_config

    def set(self, config):
        self._private_real_config = config

    def get_db_key_map(sefl, db):
        return sefl.__private_database_keys_map[db]

    def get_db_map(sefl, db):
        return sefl.__private_database_map[db]


DEFAULT_LOG_DIR = os.getcwd()
DEFAULT_LOGGING_CONFIG = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "console": {
            "class": "logging.Formatter",
            "format": (
                "[%(asctime)s] [%(levelname)s] (%(name)s) " "%(message)s (%(processName)-10s - pid: %(process)d)"
            ),
            "datefmt": "%Y-%m-%d %H:%M:%S",
        },
        "file": {
            "class": "logging.Formatter",
            "format": (
                "[%(asctime)s] [%(levelname)s] (%(name)s) " "%(message)s (%(processName)-10s - pid: %(process)d)"
            ),
            "datefmt": "%Y-%m-%d %H:%M:%S",
        },
    },
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
            "formatter": "console",
            "level": logging.INFO,
        },
        "file": {
            "class": "logging.handlers.RotatingFileHandler",
            "filename": os.path.join(DEFAULT_LOG_DIR, "planetmint.log"),
            "mode": "w",
            "maxBytes": 209715200,
            "backupCount": 5,
            "formatter": "file",
            "level": logging.INFO,
        },
        "errors": {
            "class": "logging.handlers.RotatingFileHandler",
            "filename": os.path.join(DEFAULT_LOG_DIR, "planetmint-errors.log"),
            "mode": "w",
            "maxBytes": 209715200,
            "backupCount": 5,
            "formatter": "file",
            "level": logging.ERROR,
        },
    },
    "loggers": {},
    "root": {
        "level": logging.DEBUG,
        "handlers": ["console", "file", "errors"],
    },
}
